### PR TITLE
Updated describe command to show layer's digest of each image

### LINKED ISSUE #348: `describe` command should provide information about the layers of each Image
**Describe the problem/challenge you have**
To help with more information about the Bundle, `imgpkg` should provide the layer SHA for each Image, this could help to better understand what are the moving parts of the bundle and eventually help do some debugging

**Describe the solution you'd like**
Expected output:

```
Bundle SHA: sha256:3cd8f0d917a92fafeef2d44eb63195d5351c35cf46b8b7448cb126c7d68c6291

Images:
  - Image: localhost:5000/test@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
    Type: Image
    Origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
    Layers:
        - Digest: sha256:8ece9ac45f2b7228b2ed95e9f407b4f0dc2ac74f93c62ff1156f24c53042ba54


  - Image: localhost:5000/test@sha256:510482df49db421542ae10142e8ce99572b23ef72675c37e92e4a12b541a3f6a
    Type: Internal
    Layers:
        - Digest: sha256:39498e8ae7157666d1f48b549ffd5fc8331e2a241c9fbd8a619522dc93b29c95

  - Image: localhost:5000/test@sha256:ecc10da4f480b1ba263d96197151d76bc787104cbcf7f1b1f42b14340b58bae7
    Type: Signature
    Annotations:
      tag: sha256-4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0.sig
    Layers:
        - Digest: sha256:04120f5a629bb41ef33ab4a54bddb0a74d46aa1984a86762ffea4663f8b137d8

Succeeded
```

On the yaml output it should look like:
```yaml
sha: sha256:3cd8f0d917a92fafeef2d44eb63195d5351c35cf46b8b7448cb126c7d68c6291
content:
  images:
    sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0:
      annotations:
        kbld.carvel.dev/id: index.docker.io/dkalinin/k8s-simple-app:latest
        kbld.carvel.dev/origins: |
          - resolved:
              tag: latest
              url: index.docker.io/dkalinin/k8s-simple-app:latest
      image: localhost:5000/test@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      imageType: Image
      layers:
        - digest: sha256:8ece9ac45f2b7228b2ed95e9f407b4f0dc2ac74f93c62ff1156f24c53042ba54
      origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
    sha256:510482df49db421542ae10142e8ce99572b23ef72675c37e92e4a12b541a3f6a:
      image: localhost:5000/test@sha256:510482df49db421542ae10142e8ce99572b23ef72675c37e92e4a12b541a3f6a
      imageType: Internal
      layers:
        - digest: sha256:39498e8ae7157666d1f48b549ffd5fc8331e2a241c9fbd8a619522dc93b29c95
      origin: localhost:5000/test@sha256:510482df49db421542ae10142e8ce99572b23ef72675c37e92e4a12b541a3f6a
    sha256:ecc10da4f480b1ba263d96197151d76bc787104cbcf7f1b1f42b14340b58bae7:
      annotations:
        tag: sha256-4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0.sig
      image: localhost:5000/test@sha256:ecc10da4f480b1ba263d96197151d76bc787104cbcf7f1b1f42b14340b58bae7
      imageType: Signature
      layers:
        - digest: sha256:04120f5a629bb41ef33ab4a54bddb0a74d46aa1984a86762ffea4663f8b137d8
      origin: localhost:5000/test@sha256:ecc10da4f480b1ba263d96197151d76bc787104cbcf7f1b1f42b14340b58bae7
image: localhost:5000/test@sha256:3cd8f0d917a92fafeef2d44eb63195d5351c35cf46b8b7448cb126c7d68c6291
metadata: {}
origin: localhost:5000/test@sha256:3cd8f0d917a92fafeef2d44eb63195d5351c35cf46b8b7448cb126c7d68c6291
```

**Anything else you would like to add:**
For the yaml output maybe we can add more information for the layers, like the mediaType and eventually annotations, this could be interesting since `cosign` annotates layers on the OCI Image

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help work on this issue.

### PR DESCRIPTION
**Pull request by kumaritanushree**
Fixes https://github.com/carvel-dev/imgpkg/issues/348

**After this changes output will look like:**
```
$ ./imgpkg describe -b k8slt/kc-e2e-test-repo 
Bundle SHA: sha256:ddd93b67b97c1460580ca1afd04326d16900dc716c4357cade85b83deab76f1c

Images:
  - Image: index.docker.io/k8slt/kctrl-example-pkg@sha256:73713d922b5f561c0db2a7ea5f4f6384f7d2d6289886f8400a8aaf5e8fdf134a
    Type: Bundle
    Origin: index.docker.io/k8slt/kctrl-example-pkg@sha256:73713d922b5f561c0db2a7ea5f4f6384f7d2d6289886f8400a8aaf5e8fdf134a
    Layers:
      - Digest: sha256:dc5da8ac3e06c740dc7d82f9e011ef6927ca9893b5dad7e95650f53ba89445df
    Annotations:
      kbld.carvel.dev/id: k8slt/kctrl-example-pkg:v2.0.0
    Images:
    - Image: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      Type: Image
      Origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      Layers:
        - Digest: sha256:8ece9ac45f2b7228b2ed95e9f407b4f0dc2ac74f93c62ff1156f24c53042ba54
      Annotations:
        kbld.carvel.dev/id: docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0

  - Image: index.docker.io/k8slt/kctrl-example-pkg@sha256:8ffa7f9352149dba1d539d0006b38eda357917edcdd39b82497a61dab2c27b75
    Type: Bundle
    Origin: index.docker.io/k8slt/kctrl-example-pkg@sha256:8ffa7f9352149dba1d539d0006b38eda357917edcdd39b82497a61dab2c27b75
    Layers:
      - Digest: sha256:52bf80eaaecb06062519d66bf59bc8a093cebd37faf48e47fdd79fd009d83612
    Annotations:
      kbld.carvel.dev/id: k8slt/kctrl-example-pkg:v1.0.0
    Images:
    - Image: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      Type: Image
      Origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      Layers:
        - Digest: sha256:8ece9ac45f2b7228b2ed95e9f407b4f0dc2ac74f93c62ff1156f24c53042ba54
      Annotations:
        kbld.carvel.dev/id: docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0

Succeeded
```

**In yaml format:**
```
$ ./imgpkg describe -b k8slt/kc-e2e-test-repo -oyaml
sha: sha256:ddd93b67b97c1460580ca1afd04326d16900dc716c4357cade85b83deab76f1c
content:
  bundles:
    sha256:8ffa7f9352149dba1d539d0006b38eda357917edcdd39b82497a61dab2c27b75:
      annotations:
        kbld.carvel.dev/id: k8slt/kctrl-example-pkg:v1.0.0
      content:
        images:
          sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0:
            annotations:
              kbld.carvel.dev/id: docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
            image: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
            imageType: Image
            layers:
            - digest: sha256:8ece9ac45f2b7228b2ed95e9f407b4f0dc2ac74f93c62ff1156f24c53042ba54
            origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      image: index.docker.io/k8slt/kctrl-example-pkg@sha256:8ffa7f9352149dba1d539d0006b38eda357917edcdd39b82497a61dab2c27b75
      layers:
      - digest: sha256:52bf80eaaecb06062519d66bf59bc8a093cebd37faf48e47fdd79fd009d83612
      metadata: {}
      origin: index.docker.io/k8slt/kctrl-example-pkg@sha256:8ffa7f9352149dba1d539d0006b38eda357917edcdd39b82497a61dab2c27b75
    sha256:73713d922b5f561c0db2a7ea5f4f6384f7d2d6289886f8400a8aaf5e8fdf134a:
      annotations:
        kbld.carvel.dev/id: k8slt/kctrl-example-pkg:v2.0.0
      content:
        images:
          sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0:
            annotations:
              kbld.carvel.dev/id: docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
            image: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
            imageType: Image
            layers:
            - digest: sha256:8ece9ac45f2b7228b2ed95e9f407b4f0dc2ac74f93c62ff1156f24c53042ba54
            origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      image: index.docker.io/k8slt/kctrl-example-pkg@sha256:73713d922b5f561c0db2a7ea5f4f6384f7d2d6289886f8400a8aaf5e8fdf134a
      layers:
      - digest: sha256:dc5da8ac3e06c740dc7d82f9e011ef6927ca9893b5dad7e95650f53ba89445df
      metadata: {}
      origin: index.docker.io/k8slt/kctrl-example-pkg@sha256:73713d922b5f561c0db2a7ea5f4f6384f7d2d6289886f8400a8aaf5e8fdf134a
image: index.docker.io/k8slt/kc-e2e-test-repo@sha256:ddd93b67b97c1460580ca1afd04326d16900dc716c4357cade85b83deab76f1c
layers:
- digest: sha256:a18d039a28a58cba06a4aaf435e4479b88daeb3cf7a84e1ca1a5d913f56ad1a8
metadata: {}
origin: index.docker.io/k8slt/kc-e2e-test-repo@sha256:ddd93b67b97c1460580ca1afd04326d16900dc716c4357cade85b83deab76f1c

Succeeded
```

### CODE REVIEW

File: pkg/imgpkg/v1/describe.go [hunk 1]
```diff
@@ -40,6 +43,7 @@ type ImageInfo struct {
 	Annotations map[string]string `json:"annotations,omitempty"`
 	ImageType   bundle.ImageType  `json:"imageType"`
 	Error       string            `json:"error,omitempty"`
+	Layers      []string          `json:"layers,omitempty"`
```
Comment (rohitagg2020) [inline pkg/imgpkg/v1/describe.go]: Can this be empty?

File: pkg/imgpkg/v1/describe.go [hunk 2]
```diff
@@ -189,3 +196,30 @@ func (r *refWithDescription) describeBundleRec(visitedImgs map[string]refWithDes
 
 	return desc.bundle
 }
+
+func getImageLayersDigest(image string) []string {
+	layers := []string{}
+	parsedImgRef, err := regname.ParseReference(image, regname.WeakValidation)
+	if err != nil {
+		panic(fmt.Sprintf("Error: [%s] in parsing image: %s", err.Error(), image))
```
Comment (rohitagg2020) [inline pkg/imgpkg/v1/describe.go]: Do we need to panic? Can we handle it more gracefully?

File: pkg/imgpkg/v1/describe.go (hunk 1 above)
Comment (joaopapereira) [reply pkg/imgpkg/v1/describe.go]: The only situation is if we have an error.

File: test/e2e/describe_test.go [hunk 3]
```diff
@@ -358,42 +417,56 @@ content:
         some.other.annotation: some other value
       image: %s%s
       imageType: Image
+      layers:
+      - %s
```
Comment (joaopapereira) [inline test/e2e/describe_test.go]: According to the story the output should look like 
```yaml
      image: localhost:5000/test@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
      imageType: Image
      layers:
        - digest: sha256:8ece9ac45f2b7228b2ed95e9f407b4f0dc2ac74f93c62ff1156f24c53042ba54
      origin: index.docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
```

File: test/helpers/assertion.go [hunk 4]
```diff
@@ -91,7 +91,7 @@ func (a *Assertion) ValidateImagesPresenceInRegistry(refs []string) error {
 
 func (a *Assertion) ValidateCosignSignature(refs []string) {
 	for _, ref := range refs {
-		cmdArgs := []string{"verify", "-key", filepath.Join(a.signatureKeyLocation, "cosign.pub"), ref}
+		cmdArgs := []string{"verify", "--key", filepath.Join(a.signatureKeyLocation, "cosign.pub"), ref}
```
Comment (joaopapereira) [inline test/helpers/assertion.go]: Is this change related to this PR? If not let us at least isolate this into a separated commit

File: pkg/imgpkg/v1/describe.go (hunk 2 above)
Comment (joaopapereira) [reply pkg/imgpkg/v1/describe.go]: Since this is an API, we should be careful when panicking. We should always give control back to the caller. In the case of this function, I think we should not panic and return the error.

Comment (kumaritanushree) [reply pkg/imgpkg/v1/describe.go]: `DescribeBundle` is not returning any error hence we have to use panic to show user error and exit from this. 
@joaopapereira is there any specific reason to not to return error from `DescribeBundle`  [here](https://github.com/carvel-dev/imgpkg/blob/develop/pkg/imgpkg/v1/describe.go#L109) ?

Comment (joaopapereira) [reply pkg/imgpkg/v1/describe.go]: In the past we did not need to, if we need to return errors now we should do it.

**Review by joaopapereira — CHANGES_REQUESTED** [review]
Added some comments in the PR

File: test/helpers/assertion.go (hunk 4 above)
Comment (kumaritanushree) [reply test/helpers/assertion.go]: not required, removed

File: test/e2e/describe_test.go (hunk 3 above)
Comment (kumaritanushree) [reply test/e2e/describe_test.go:421]: In text output it will look same while yaml output will have what there in e2e test. Yaml format is not we are controlling manually. This one we get when we do marshalling of struct https://github.com/carvel-dev/imgpkg/blob/develop/pkg/imgpkg/cmd/describe.go#L194.

File: pkg/imgpkg/v1/describe.go [hunk 5]
```diff
@@ -156,27 +167,36 @@ func (r *refWithDescription) describeBundleRec(visitedImgs map[string]refWithDes
 
 	for _, ref := range imagesRefs {
 		if ref.IsBundle == nil {
-			panic("Internal consistency: IsBundle after processing must always have a value")
+			return desc.bundle, fmt.Errorf("Internal consistency: IsBundle after processing must always have a value")
```
Comment (praveenrewar) [inline pkg/imgpkg/v1/describe.go]:
```suggestion
			return desc.bundle, fmt.Errorf("Internal inconsistency: IsBundle after processing must always have a value")
```

File: pkg/imgpkg/v1/describe.go [hunk 6]
```diff
@@ -146,7 +157,7 @@ func (r *refWithDescription) describeBundleRec(visitedImgs map[string]refWithDes
 		}
 	}
 	if newBundle == nil {
-		panic(fmt.Sprintf("Internal consistency: bundle with ref '%s' could not be found in list of bundles", currentBundle.PrimaryLocation()))
+		return desc.bundle, fmt.Errorf("Internal consistency: bundle with ref '%s' could not be found in list of bundles", currentBundle.PrimaryLocation())
```
Comment (praveenrewar) [inline pkg/imgpkg/v1/describe.go]:
```suggestion
		return desc.bundle, fmt.Errorf("Internal inconsistency: bundle with ref '%s' could not be found in list of bundles", currentBundle.PrimaryLocation())
```

File: pkg/imgpkg/v1/describe.go [hunk 7]
```diff
@@ -187,5 +207,32 @@ func (r *refWithDescription) describeBundleRec(visitedImgs map[string]refWithDes
...
 
-	return desc.bundle
+	return desc.bundle, nil
+}
+
+func getImageLayersDigest(image string) ([]string, error) {
+	layers := []string{}
+	parsedImgRef, err := regname.ParseReference(image, regname.WeakValidation)
+	if err != nil {
+		return nil, err
```
Comment (praveenrewar) [inline pkg/imgpkg/v1/describe.go]: Should we format the error to add a small message, like "Parsing image reference: _err_"? This will help in knowing where the error originated if at all someone gets an error here.
Same for places below.

File: test/e2e/describe_test.go (hunk 3 above)
Comment (kumaritanushree) [reply test/e2e/describe_test.go]: updated

File: pkg/imgpkg/v1/describe.go [hunk 8]
```diff
@@ -187,5 +212,32 @@ func (r *refWithDescription) describeBundleRec(visitedImgs map[string]refWithDes
...
 
-	return desc.bundle
+	return desc.bundle, nil
+}
+
+func getImageLayersInfo(image string) ([]Layers, error) {
+	layers := []Layers{}
+	parsedImgRef, err := regname.ParseReference(image, regname.WeakValidation)
+	if err != nil {
+		return nil, fmt.Errorf("Error: %s in parsing image %s", err.Error(), image)
```
Comment (praveenrewar) [inline pkg/imgpkg/v1/describe.go:222]: nit: I think the error would already contain "Error", so just "Parsing image: err" something like this should be good?

**Review by praveenrewar — APPROVED** [review]
Apart from the one small nit, looks good to me.

File: pkg/imgpkg/v1/describe.go (hunk 8 above)
Comment (kumaritanushree) [reply pkg/imgpkg/v1/describe.go:222]: Seems error returned from  [ParseReference](https://github.com/google/go-containerregistry/blob/main/pkg/name/ref.go#L47) does not contain error as substring. I think this should be fine?

File: test/helpers/image_factory.go [hunk 9]
```diff
@@ -42,6 +42,25 @@ func (i *ImageFactory) ImageDigest(imgRef string) string {
 	return digest.String()
 }
 
+// GetImageLayerDigest will return image's layer digest
+func (i *ImageFactory) GetImageLayerDigest(image string) []string {
```
Comment (joaopapereira) [inline test/helpers/image_factory.go]:
```suggestion
func (i *ImageFactory) GetFirstLayerDigest(image string) []string {
```

**Review by joaopapereira — APPROVED** [review]
After you do the change I proposed in the PR, if all test pass you can merge the PR

File: test/helpers/image_factory.go (hunk 9 above)
Comment (kumaritanushree) [reply test/helpers/image_factory.go]: Now we are returning all layers hence will rename it to `GetImageLayersDigest`. Also image which is used in e2e test has single layers so should work with reading first element of the returned array.

**Review by joaopapereira — APPROVED** [review]
LGTM
